### PR TITLE
Created a debug output mode that dumps the the exact inputs/outputs of the PairHMM to a file

### LINKED ISSUE #7647: Add option to dump PairHMM debugging outputs to a file in a standard format
There has been a request to do some more work on the HMM again and its become clear that there is not an efficient way to rapidly generate large amounts of test data based on the old HMM results. It would be helpful to add an option to dump the hmm scores out to the command line in an easily machine parseable format. Here is an example of how it has been done in the past (and probably how we should do it this time): https://github.com/Intel-HLS/GKL/blob/master/src/test/resources/pairhmm-testdata.txt

### PR DESCRIPTION
**Pull request by jamesemery**
Resolves #7647

### CODE REVIEW

**Comment by jamesemery** [comment]
@droazen could you take a look at this branch?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 1]
```diff
@@ -1606,4 +1607,43 @@ private static boolean isGVCFReferenceBlock( final VariantContext vc ) {
                vc.getAlternateAlleles().size() == 1 &&
                vc.getAlternateAllele(0).equals(Allele.NON_REF_ALLELE);
     }
+
+
+    @DataProvider(name="PairHMMResultsModes")
+    public Object[][] PairHMMResultsModes() {
+        return new Object[][] {
+                //{PairHMM.Implementation.AVX_LOGLESS_CACHING, new File(TEST_FILES_DIR, "expected.AVX.hmmresults.txt")}, //TODO disabled because of failures on travis
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java:1615]: Can you elaborate on why this is disabled? Ideally we want test coverage for the default HMM in GATK...

**Review by droazen — CHANGES_REQUESTED** [review]
@jamesemery Back to you with a couple of quick comments. Main question is why the AVX_LOGLESS_CACHING test case is disabled

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/LikelihoodEngineArgumentCollection.java [hunk 2]
```diff
@@ -104,6 +104,14 @@
     @Argument(fullName="dynamic-read-disqualification-threshold", doc="Constant used to scale the dynamic read disqualificaiton")
     public double readDisqualificationThresholdConstant = PairHMMLikelihoodCalculationEngine.DEFAULT_DYNAMIC_DISQUALIFICATION_SCALE_FACTOR;
 
+    /**
+     * Argument for generating a file of all of the inputs and ouptus for the pair hmm
+     */
+    @Advanced
+    @Hidden
+    @Argument(fullName="pair-hmm-results-file", doc="File to write exact pairHMM inputs/outputs to for debugging purposes", optional = true)
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/LikelihoodEngineArgumentCollection.java:112]: all of them as far as i've tested. I have a test for all of them but the AVX was producing some different output on travis...

File: src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java [hunk 3]
```diff
@@ -228,6 +234,7 @@ public void computeLog10Likelihoods(final LikelihoodMatrix<GATKRead, Haplotype>
                         readBases, readQuals, readInsQuals, readDelQuals, overallGCP, isFirstHaplotype, nextAlleleBases);
                 logLikelihoods.set(a, readIndex, lk);
                 mLogLikelihoodArray[idx++] = lk;
+                writeToResultsFileIfApplicable(readBases, readQuals, readInsQuals, readDelQuals, overallGCP, alleleBases, lk);
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java:237]: yes i have tests for both.

File: src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java [hunk 4]
```diff
@@ -350,12 +358,46 @@ public static int findFirstPositionWhereHaplotypesDiffer(final byte[] haplotype1
...
+    }
+
+    /**
+     * Method to be invoked implementing HMM engines to be used for storing results for various debugging purposes
+     */
+    protected void writeToResultsFileIfApplicable(byte[] readBases, byte[] readQuals, byte[] readInsQuals, byte[] readDelQuals, byte[] overallGCP, byte[] alleleBases, double lk) {
+
+        if (debugOutputStream!= null) {
+            try {
+                debugOutputStream.write("\n" + new String(alleleBases) + " " + new String(readBases) + " " + SAMUtils.phredToFastq(readQuals) + " " + SAMUtils.phredToFastq(readInsQuals) + " " + SAMUtils.phredToFastq(readDelQuals) + " " + SAMUtils.phredToFastq(overallGCP) + " " + String.format("%e",lk));
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java:380]: I don't think so. If the alleles or the phred to fastq mehtods are returning results with whitespaces in them we have much bigger problems...

File: src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java [hunk 5]
```diff
@@ -350,12 +358,46 @@ public static int findFirstPositionWhereHaplotypesDiffer(final byte[] haplotype1
...
+
     /**
      * Called at the end of the program to close files, print profiling information etc 
      */
     @Override
     public void close() {
-        if(doProfiling)
-            logger.info("Total compute time in PairHMM computeLogLikelihoods() : "+(pairHMMComputeTime*1e-9));
+        if(doProfiling) {
+            logger.info("Total compute time in PairHMM computeLogLikelihoods() : " + (pairHMMComputeTime * 1e-9));
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/pairhmm/PairHMM.java:393]: seconds I believe just as it has in the past for gatk. Its using nanoTime so converting to seconds should be 10^-9.

**Comment by jamesemery** [comment]
@droazen I loosened the failing test and responded to your comments? Anything else before we can get this in?

**Review by droazen — APPROVED** [review]
👍

**Pull request merged**